src/libpriv/rpmostree-core.c

```diff
@@ -2904,10 +2935,15 @@ checkout_package (OstreeRepo   *repo,
   /* Used in the no-rofiles-fuse path */
   opts.force_copy_zerosized = force_copy_zerosized;
 
-  if (files_skip && g_hash_table_size (files_skip) > 0)
+  /* If called by `checkout_package_into_root()`, there may be files that need to be filtered */
+  if (filter_data)
     {
-      opts.filter = checkout_filter;
-      opts.filter_user_data = files_skip;
+      if ((filter_data->files_skip && g_hash_table_size (filter_data->files_skip) > 0) ||
+          g_strv_length (filter_data->files_remove_regex) > 0)
```

Comment:
I think it would be clearer if the `FilterData` struct was handled here, right where it's needed. So e.g. `checkout_package_into_root` would pass both `files_skip` and `files_remove_regex`, and here we'd unconditionally attach a filter.

The filter then would also need to check for `files_remove_regex` like it does for `files_skip`.

Reply:
Not sure if I'm understanding correctly. Currently, the filter checks for both `files_remove_regex` and `files_skip` to determine which one (possibly both) of the fields in the `FilterData` struct are needed. 
Are you suggesting that `checkout_package()` always provide attach a filter (i.e. not have the if statement in lines 2941-2942)? I was thinking that it might be helpful for `ostree_repo_checkout_at()` to not need to go through a filter (that could potentially not be filtering anything) for every single package; that's why there are the kind of duplicate if statements in both `checkout_filter()` and `checkout_package()`.

Reply:
Yeah, I think passing the filter only when needed to `ostree_repo_checkout_at()` is nice. Though what I mean is it feels odd to pass a `FilterData` to this function only to have it be inspected and passed along to libostree. Since it's more of an implementation detail of the checkout, it feels cleaner to have it assemble a `FilterData` if needed itself.

What you have right now WFM too though!

Reply:
ahh, gotcha. Yeah, your suggestion would make a lot more sense.